contracts/gas/GasEstimate.sol

```diff
@@ -0,0 +1,143 @@
+        bytes calldata payload,
+        uint256, /* l1GasPrice */
+        uint256 l1ToL2BaseFee
+    ) internal pure returns (uint256) {
+        if (keccak256(bytes(destinationChain)) == keccak256(bytes('optimism'))) {
+            return optimismL1ToL2Fee(payload, l1ToL2BaseFee);
+        }
+
+        revert UnsupportedL2Estimate(destinationChain);
+    }
+
+    /**
+     * @notice Computes the L1 to L2 fee for a contract call on the Optimism chain.
+     * @param payload The payload of the contract call
+     * @param l1ToL2BaseFee The base fee for L1 to L2
+     * @return l1DataFee The L1 to L2 data fee
+     */
+    function optimismL1ToL2Fee(bytes calldata payload, uint256 l1ToL2BaseFee)
+        internal
+        pure
+        returns (uint256 l1DataFee)
+    {
+        /* Optimism Ecotone gas model https://docs.optimism.io/stack/transactions/fees#ecotone
+             tx_compressed_size = ((count_zero_bytes(tx_data) * 4 + count_non_zero_bytes(tx_data) * 16)) / 16
+             weighted_gas_price = 16 * base_fee_scalar*base_fee + blob_base_fee_scalar * blob_base_fee
+             l1_data_fee = tx_compressed_size * weighted_gas_price
+
+           Reference implementation:
+             https://github.com/ethereum-optimism/optimism/blob/876e16ad04968f0bb641eb76f98eb77e7e1a3e16/packages/contracts-bedrock/src/L2/GasPriceOracle.sol#L138
+        */
+
+        // The new base_fee_scalar is using old dynamic_overhead_multiplier value which is currently 0.684
+        // We are setting it to un upper bound of 0.7 to encounter for future fluctuations
+        uint256 scalarPrecision = 10**6;
+        uint256 baseFeeScalar = 7 * 10**5; // 7e5 : 1e6 = 0.7
+
+        // The blob_base_fee_scalar is currently set to 0. The blob gas model is being adopted by OP validators.
+        // https://eips.ethereum.org/EIPS/eip-4844
+        uint256 blobBaseFeeScalar = 0 * scalarPrecision;
+        uint256 blobBaseFee = 0;
+
+        // Calculating transaction size in bytes that will later be divided by 16 to compress the size
+        // 68 bytes for the TX RLP encoding overhead
+        uint256 txSize = 68 * 16;
+        // GMP executeWithToken call parameters
+        // 32 bytes for the commandId, 96 bytes for the sourceChain, 128 bytes for the sourceAddress, 96 bytes for token symbol, 32 bytes for amount
+        // Expecting half of the calldata bytes to be zeroes. So multiplying by 10 as an average of 4 and 16
+        txSize += (32 + 96 + 128 + 96 + 32) * 10;
+
+        for (uint256 i; i < payload.length; ++i) {
+            if (payload[i] == 0) {
+                txSize += 4; // 4 for each zero byte
+            } else {
+                txSize += 16; // 16 for each non-zero byte
+            }
+        }
+
+        uint256 weightedGasPrice = 16 * baseFeeScalar * l1ToL2BaseFee + blobBaseFeeScalar * blobBaseFee;
+
+        l1DataFee = (weightedGasPrice * txSize) / (16 * scalarPrecision); // 16 for txSize compression and scalar precision conversion
```

Comment:
this gas estimate is in terms of L2 gas price. But it needs to be in terms of the chain's token price that the gas service is on, which is why we use relativeGasPrice instead